usbd: device_next: cdc: wait IN EP ready before enqueue data to EP

zeonchew opened the pull request:
Under the current structure of usbd_cdc_acm, there is a possibility that TX data is enqueued to UDC driver before the previous transaction is completed when host is sending data to device faster than it receives data from device, causing dropped data. 

Example of such flow is that: 
1. USB_OUT transaction completed. Data passed to sample's application code. Applicaiton code echo the data back into `tx_fifo`. 
2. USB_IN transaction is enqueued to UDC driver. 
3. Before USB_IN transaction is completed, another USB_OUT transaction is completed. Application code echo the data back into `tx_fifo` again. 
4. Since the echo raised `tx_fifo.altered` flag, `tx_fifo_work` is queued again. 
5. When `cdc_acm_tx_fifo_handler()` tries to enqueue packet to UDC driver, it is rejected because the USB_IN endpoint is still busy, causing the packet to be dropped. 

This PR adds mechanism to wait for IN endpoint to complete existing transaction before allowing subsequent transaction to be enqueued to UDC drive.

zeonchew committed:
usbd: device_next: cdc: wait IN EP ready before enqueue data to EP

Under the current structure of usbd_cdc_acm, there is a possibility that
TX data is enqueued to UDC driver before the previous transaction is
completed, hence causing dropped data. This commit adds a mechanism for
usbd_cdc_acm driver to wait for IN endpoint to complete existing
transaction before subsequent transaction is enqueued. This commit also
resolves issue whee altered flag is cleared after a tx transaction even
if there is remaining data in tx_fifo, by checking for the fifo empty
status directly instead.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -540,6 +544,11 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	}
 
+	if (atomic_test_and_set_bit(&data->state, CDC_ACM_TX_EP_BUSY)) {
+		cdc_acm_work_submit(&data->tx_fifo_work);
```
How does this work? Will it simply keep retriggering the TX fifo work until the request is completed? This sounds way too much CPU intensive to me that could starve system workqueue and other lower priority threads.

zeonchew replied:
Hi @tmon-nordic, you are right. Indeed keep requeuing tx_fifo_work would cause the issue you mentioned, I have totally missed that. I have changed the handling to raised the tx_fifo altered flag instead. Please help and see whether it make sense for you.

tmon-nordic commented:
The commit message should not be a concatenation of squashed commit messages like it is now.

zeonchew commented:
> The commit message should not be a concatenation of squashed commit messages like it is now.

Hi @tmon-nordic, thanks again. Sorry that I am unaware for the rules and workflow. Have removed the "review history" from the commit message. 

May I check with you whether there is documentation regarding correct workflow for addressing review? 

From what I understand from what is happening so far is that, for addressing review, we wanted the amendment to be squashed with the original commit, without needing record of the review history. 
Do I understand this correctly? 

Thanks again @tmon-nordic.

tmon-nordic commented:
> May I check with you whether there is documentation regarding correct workflow for addressing review?

It can be derived from https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html but only indirectly. The PR in Zephyr is essentially equivalent to what a patch series in Linux Kernel is.

> From what I understand from what is happening so far is that, for addressing review, we wanted the amendment to be squashed with the original commit, without needing record of the review history. Do I understand this correctly?

The review history is not really useful in the "final patch set" (each subsequent force push should be possible to consider as "final patch set", there shouldn't really be "WIP" changes unless you mark the PR as draft).

There are so many commits that if all review history was in the main history it would be orders of magnitude harder to navigate it. If someone is interested in review history then the GitHub is the place to look for it (Linux Kernel IMHO has it better with the lore.kernel.org archive, but that's related to tooling and not necessarily the development process).

zeonchew commented:
@tmon-nordic, undestood. Thanks for the sharing again 😄

jfischer-no reviewed (changes requested):
>Under the current structure of usbd_cdc_acm, there is a possibility that
TX data is enqueued to UDC driver before the previous transaction is
completed, hence causing the dropped data.

Where is data dropped? Also please take a look at #75439

zeonchew commented:
Hi @jfischer-no, 

> Where is data dropped?

from my current understanding, because that `cdc_acm_irq_cb_handler()` is invoked-able by a number of events, there is a possibility that when the handler is invoked before TX is completed, and `tx_fifo.altered` is raised. in this case `usbd_cdc_acm` would still enqueue the data from `tx_fifo` to UDC driver at `cdc_acm_tx_fifo_handler()` by calling `usbd_ep_enqueue()`. Since the EP status is still busy at UDC driver side, depending on the enqueue mechanism at UDC, it would be either one of the following scenario: 
1. <async implementation> UDC driver push the request into a queue and returns success, but when the UDC thread tries to execute the TX transaction, it fails and discarded by UDC driver. 
2. <sync implementation> UDC driver return failure for the calls, and the failure case handling in `cdc_acm_tx_fifo_handler()` discards the buffer: 

```
	ret = usbd_ep_enqueue(c_data, buf);
	if (ret) {
		LOG_ERR("Failed to enqueue");
		net_buf_unref(buf);
	}
```

> Also please take a look at #75439

From my understanding of the code change in #75439, it seems that the `k_work_delayable` is very useful to delay the `tx_fifo_work` when the EP is busy. It would be good if we could use delayed `tx_fifo_work` to retrigger the work instead of setting `data->tx_fifo.altered = true` as what is done in this PR. What do you think?

By the way, it seems that the case where "IN endpoint is busy" is not included in #75439 yet. Please correct me if I understand that wrongly.

tmon-nordic commented:
> From my understanding of the code change in #75439, it seems that the `k_work_delayable` is very useful to delay the `tx_fifo_work` when the EP is busy. It would be good if we could use delayed `tx_fifo_work` to retrigger the work instead of setting `data->tx_fifo.altered = true` as what is done in this PR. What do you think?

I think blindly retriggering with some arbitrary delay is bad design. It should really be a flag that executes once the underlying condition has changed (which is the case here with `data->tx_fifo.altered`.

The two issues seem independent to me. But it would be best @zeonchew could try to trigger the issue this PR is aiming to fix with #75439 changes applied (and without this PR).

zeonchew commented:
> > From my understanding of the code change in #75439, it seems that the `k_work_delayable` is very useful to delay the `tx_fifo_work` when the EP is busy. It would be good if we could use delayed `tx_fifo_work` to retrigger the work instead of setting `data->tx_fifo.altered = true` as what is done in this PR. What do you think?
> 
> I think blindly retriggering with some arbitrary delay is bad design. It should really be a flag that executes once the underlying condition has changed (which is the case here with `data->tx_fifo.altered`).
> 
> The two issues seem independent to me. But it would be best @zeonchew could try to trigger the issue this PR is aiming to fix with #75439 changes applied (and without this PR).

@tmon-nordic, sure. Let me have a try on it soonest possible and get back here to update the status. 😉

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -114,6 +115,8 @@ struct cdc_acm_uart_data {
 	atomic_t state;
 	struct k_sem notif_sem;
+	/* USBD CDC ACM TX transaction deferred due to BULK IN EP busy */
+	bool tx_deferred;
```
Why new variable instead of just using `data->tx_fifo.altered` like it was earlier?

zeonchew replied:
Hi @tmon-nordic, 

I was just about to add my report on what I have done here. 

Since that #75439 has been merged into zephyr main, I have rebased my works and retested with the changes of #75439 applied. However, I am still able to reproduce this issue. 

I have also run more stress test on my fixed ported over to the latest main, and found out that raising `data->tx_fifo.altered` wouldn't help in retrigger the TX transaction, as `tx_fifo.altered` is cleared in `cdc_acm_irq_cb_handler()`. 

https://github.com/zephyrproject-rtos/zephyr/blob/c710f8892beef3e486dded2f0ad4394e4a4d9aa0/subsys/usb/device_next/class/usbd_cdc_acm.c#L812C1-L813C32

The previous testing wasn't rigorous enough, that the subsequent transaction was actually triggered by new data coming in from RX instead (and hence the stale data was flushed out together). 

This being said, I have added another handling here to make sure that remaining data in tx_fifo, when the endpoint is unable to send it out in one shot, gets sent out immediately after current TX is completed as well. 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L571C2-L580C3

Please help to review whether this make sense.

tmon-nordic replied:
I don't understand `cdc_acm_irq_cb_handler()`. Isn't the `data->tx_fifo.altered = false;` going to make the `if (data->tx_fifo.altered)` effectively a dead code? Shouldn't this be fixed instead of adding new boolean?

zeonchew replied:
> I don't understand `cdc_acm_irq_cb_handler()`. Isn't the `data->tx_fifo.altered = false;` going to make the `if (data->tx_fifo.altered)` effectively a dead code? Shouldn't this be fixed instead of adding new boolean?

I took me a while to understand this as well, from my understanding, in this handler, it will first clear all the altered flag, and then callback to the interrupt handler at application side: `data->cb(...)`. 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L841-L844

If there is more data to be transmitted by the application, the `tx_fifo.altered` flag will be raised by application before the callback returns. Then here we go, continue to send data out again. 

I see that this part is checked in by @jfischer-no. @jfischer-no, can you please help to confirm whether this understanding is correct? 
Thanks.

tmon-nordic replied:
Isn't there a problem when there is more data in the tx ringbuffer? The sequence I think about is:
  * `cdc_acm_irq_cb_handler()` sets `data->tx_fifo.altered = false;`
  * `data->cb(usbd_class_get_private(c_data), data->cb_data);` is called, application writes to ring buffer more bytes than a CDC ACM TX buffer can fit
  * TX fifo work is triggered because `data->tx_fifo.altered` was set to true alongside ring buffer write
  * `cdc_acm_tx_fifo_handler()` calls `len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);` and submits the transfer. The ring buffer still has pending TX data.
  * host reads all the data, `usbd_cdc_acm_request()` submits `cdc_acm_work_submit(&data->irq_cb_work);`
  * `cdc_acm_irq_cb_handler()` sets `data->tx_fifo.altered = false;`
  * `data->cb(usbd_class_get_private(c_data), data->cb_data);` is called, application doesn't have more data to write
  * TX fifo work is not triggered despite there being data in the ring buffer.

Is there anything preventing the above scenario?

zeonchew replied:
Hi @tmon-nordic, 

I added this for your case you mentioned in the latest commit just now ☺️ 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L571-L580

tmon-nordic replied:
Why to have two booleans for roughly the same? Wouldn't it be better to change the `data->tx_fifo.altered` logic completely? If we go with separate variable, I think `bool tx_deferred;` should be moved inside `tx_fifo` structure.

But is it really necessary to have `tx_deferred`? I think that the `CDC_ACM_TX_EP_BUSY` in combination with `ring_buf_is_empty(data->tx_fifo.rb))` should be enough to determine whether the TX work has to be raised or not.

My idea would be to replace `if (data->tx_fifo.altered)` with `if (!ring_buf_is_empty(data->tx_fifo.rb)) && !atomic_test_and_set_bit(&data->state, CDC_ACM_TX_EP_BUSY))`. Then the `CDC_ACM_TX_EP_BUSY` would be cleared in `usbd_cdc_acm_request()` when the IN request completes. This would essentially eliminate the need for `data->tx_fifo.altered` altogether and potentially make the code simpler to follow and more robust. Am I missing something?

zeonchew replied:
Hi @tmon-nordic, I am new to open source development and zephyr OS as well. Wanted to check with you, is that okay to do such a change without consulting the original author?

tmon-nordic replied:
Open Source development works on technical merit. I would say that there is no need to consult the original author when proposing a change that solves issues or otherwise enhances the code base.

While there may be times where you want to ask what was intended, there is also the possibility that the original author simply is no longer around.

zeonchew replied:
sorry, was going to replace my comment with another one just now but somehow I lost all of them. 
(The comment on saying I am new to Open-Source and ZephyrOS, asking on how this community works)

I am kind of agree that in this case, checking for `ring_buf_is_empty()` at  `cdc_acm_irq_cb_handler()`  plus checking for `CDC_ACM_TX_EP_BUSY` flag at `cdc_acm_tx_fifo_handler()` would be optimal as checking the ring buffer status directly gives us the most accurate status for actions to be taken. 

It seems that the original intend of `altered` flag is to provide a fast access to the status whether application layer has accessed the ring buffer. However, it seems to me that the bit might not be good enough for the scenario in tx_fifo here anymore. 

@jfischer-no , what do you think? Is @tmon-nordic suggestion here to check whether to trigger `tx_fifo_work` based on whether `ring_buf_is_empty()` sounds good to you as well?

zeonchew replied:
Hi @tmon-nordic, changes has been made as discussed and tested working at my side. 
Please kindly review and see whether its all good for you. 
Thanks.

SynchronicIT commented:
This is probably a duplicate of #78160. The same changes for a different reason have been made.

zeonchew commented:
> This is probably a duplicate of #78160. The same changes for a different reason have been made. 

Hi @SynchronicIT, indeed it looks like both PR are aiming to resolve the same issue  however, you could read from the discussion above, we have came out with some improvement here to make the code more efficient.

zeonchew commented:
@jfischer-no, do the final change looks good for you? After discussion with @tmon-nordic, we believe current change is the best solution for now. If it matches your expectation, perhaps we might want to merge it in. Thanks.

jfischer-no commented:
> Hi @jfischer-no,
> 
> > Where is data dropped?
> 
> from my current understanding, because that `cdc_acm_irq_cb_handler()` is invoked-able by a number of events, there is a possibility that when the handler is invoked before TX is completed, and `tx_fifo.altered` is raised. in this case `usbd_cdc_acm` would still enqueue the data from `tx_fifo` to UDC driver at `cdc_acm_tx_fifo_handler()` by calling `usbd_ep_enqueue()`. Since the EP status is still busy at UDC driver side, depending on the enqueue mechanism at UDC, it would be either one of the following scenario:
> 
>     1. UDC driver push the request into a queue and returns success, but when the UDC thread tries to execute the TX transaction, it fails and discarded by UDC driver.
> 
>     2. UDC driver return failure for the calls, and the failure case handling in `cdc_acm_tx_fifo_handler()` discards the buffer:

For 1), If the driver fails to start a new transfer, the driver would return the buffer to a higher level with an error. For 2), very unlikely, but then there is nothing we can do and your patch does not change it.

On which platform in the tree are you observing the problems described in this PR?

jfischer-no reviewed (changes requested):
>Under the current structure of usbd_cdc_acm, there is a possibility that
TX data is enqueued to UDC driver before the previous transaction is
completed, hence causing dropped data.

This is not true, it is fine to enqueue multiple transfers. Here, the number of transfers is limited, but it does not cause any data to be dropped.

Also altered flag in tx_fifo should not be removed.

zeonchew commented:
> > Hi @jfischer-no,
> > > Where is data dropped?
> > 
> > 
> > from my current understanding, because that `cdc_acm_irq_cb_handler()` is invoked-able by a number of events, there is a possibility that when the handler is invoked before TX is completed, and `tx_fifo.altered` is raised. in this case `usbd_cdc_acm` would still enqueue the data from `tx_fifo` to UDC driver at `cdc_acm_tx_fifo_handler()` by calling `usbd_ep_enqueue()`. Since the EP status is still busy at UDC driver side, depending on the enqueue mechanism at UDC, it would be either one of the following scenario:
> > ```
> > 1. UDC driver push the request into a queue and returns success, but when the UDC thread tries to execute the TX transaction, it fails and discarded by UDC driver.
> > 
> > 2. UDC driver return failure for the calls, and the failure case handling in `cdc_acm_tx_fifo_handler()` discards the buffer:
> > ```
> 
> For 1), If the driver fails to start a new transfer, the driver would return the buffer to a higher level with an error. For 2), very unlikely, but then there is nothing we can do and your patch does not change it.
> 
> On which platform in the tree are you observing the problems described in this PR?

Hi @jfischer-no, 

For 1), there is no way the driver would be able to return the failure status since that the call has returned success when it was queued, and this is the handling from `udc_nrf.c`, it would just return because the endpoint is still busy. 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/drivers/usb/udc/udc_nrf.c#L96-L102

Please have a look again at the code change because this PR does exactly what it takes to resolve the issue. It marks that the endpoint is busy when TX data is enqueued to the endpoint here: 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L553-L555

and then it marks it as idle when Endpoint completed the transfer: 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L249-L255

In any case that the endpoint is busy when `cdc_acm_tx_fifo_handler()` trying to enqueue the data to endpoint. It is returned so that the handling is deferred to the next invocation of `cdc_acm_irq_cb_handler()` by the completion of the TX endpoint, which will retrigger the enqueuing of the remaining data in tx_fifo to the endpoint: 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L831-L834

We are making sure that we do not even try to enqueue data to TX endpoint when it is busy so that the error will not even have a chance to occur here.

zeonchew commented:
> > Under the current structure of usbd_cdc_acm, there is a possibility that
> > TX data is enqueued to UDC driver before the previous transaction is
> > completed, hence causing dropped data.
> 
> This is not true, it is fine to enqueue multiple transfers. Here, the number of transfers is limited, but it does not cause any data to be dropped.
> 
> Also altered flag in tx_fifo should not be removed.

As pointed out by the previous response, since that the data has been popped out from the tx_fifo before it is enqueued to the endpoint. The data is now not available in tx_fifo anymore. 

For case (1) where UDC driver return failure outright by the udc driver, current handling is to just unref the buffer, which means that the data is dropped: 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L566-L570

For case (2) where UDC driver enqueues the transaction by returning success and discard the event later due to endpoint busy, there is no chance for usbd_cdc_acm to recover the data at all. 

Hence yes, trying to enqueuing data to TX endpoint does cause data drop, and this PR avoid that.

The reason for replacing `tx_fifo.altered` with direct check on `tx_fifo` empty status, is that due to the fact that `tx_fifo.altered` is cleared for every `cdc_acm_irq_cb_handler()` invocation, which in the following case, will not indicate accurately whether there is still data to be transmitted: 
- application code push in more data that the TX endpoint could transmit in a single data->cb(). In this case, `tx_fifo.altered` will be cleared in `cdc_acm_irq_cb_handler()`, and not raised by `data->cb()`, because all the data was pushed in by the application code through previuous `data->cb()`. Hence, the 2nd part of the data that is supposed to be sent is then not enqueued, until there is new event that could invoke `tx_fifo_work()` occurs. 
- related to this PR: Condition such as EP is still busy when `tx_fifo_work` tries to enqueue data when endpoint is still busy, will have the enqueue delayed, which will again, cause no invocation of `tx_fifo_work()` when tx endpoint complete event occurs, causing data not enqueued to endpoint when it is supposed to.

SynchronicIT commented:
I would propose to mark this pull request as duplicate of #78160. 
- In case the EP enqueue fails, this patch has the possibility for a deadlock. (flag busy remains set)
- the preferred route is to remain the tx_fifo.altered approach as it currently is

zeonchew commented:
> I would propose to mark this pull request as duplicate of #78160.
> 
> * In case the EP enqueue fails, this patch has the possibility for a deadlock. (flag busy remains set)
> * the preferred route is to remain the tx_fifo.altered approach as it currently is

1. Can you point out on the possibility of deadlock? 
2. In here we are proposing to not use tx_fifo.altered due to the possible cases where data written withheld in the ringbuffer. 

Still, it is trying to address the same issue, but with different approach. 

I am new to open source @tmon-nordic, are you able to advice what should we do in this case? It seems that the other PR was created earlier, but I am in favor of the solution discussed here.

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -546,6 +550,10 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 		return;
 	}
 
+	if (atomic_test_and_set_bit(&data->state, CDC_ACM_TX_EP_BUSY)) {
```
Shouldn't this be moved below the `if (buf == NULL)` check? Isn't this the deadlock that @SynchronicIT is referring to?

zeonchew replied:
I think what @SynchronicIT meant applies to both 
https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L557-L561 

and 

https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/subsys/usb/device_next/class/usbd_cdc_acm.c#L566-L570

where in both cases, the TX_EP_BUSY status will be stuck at busy. To resolve this, we could add release for the both if. IMO, putting the `atomic_test_and_set_bit()` any further might not be what we wanted, as we would not want to allocate buffer before we know that the endpoint is ready for transmission. It looks like it's "cheaper" to set the bit first, and then clear it when either of the failure case occurs. 

@tmon-nordic , should we make this ticket as duplicate like what @SynchronicIT suggested, since his PR is created earlier? I am still in favor of this solution, but I am not sure whether that is the correct thing to do in Zephyr community? 

Thank you.

tmon-nordic replied:
If you add the clears then how does this PR differ from @SynchronicIT?

zeonchew replied:
Comparing the 2 PR, What I can see are these: 
1. [Similar] The flag name, and the places to set and clear the BUSY flag are the same. (after adding clears accordingly)
2. [Different] When RX complete event occurs, @SynchronicIT PR queues an additional `tx_fifo_work` directly from `usbd_cdc_acm_request()` if tx_fifo is not empty (on top of the existing `irq_cb_work`), while for my PR, we are doing the equal handling at `cdc_acm_irq_cb_handler()`. 
3. [Different] @SynchronicIT PR is using `tx_fifo.altered` to indicate new transfer, and tx_fifo empty status as indication of new transaction requested. Which my PR use only tx_fifo empty status. Although that @SynchronicIT  PR will spawn 2 work on TX complete interrupt (`irq_cb_work` + `tx_fifo_work`), the new handling in this PR itself should resolve the conflict by deferring the older one, hence should not be an issue. 

I think I somehow missed out point (2), hence thinking it would not resolve the possible cases where data written withheld in the ringbuffer issue. 
So for now, I do think that #78160, although it is not totally as what I would love the change to be, I think it resolves the issues I am trying tor resolve.

zeonchew replied:
@tmon-nordic, if I understands correctly, since it seems that #78160 is already capable of fixing all the issues we foresee, this is where I add "duplicate of #78160" comment, and close this PR right?

jfischer-no commented:
> > > Hi @jfischer-no,
> > > > Where is data dropped?
> > > 
> > > 
> > > from my current understanding, because that `cdc_acm_irq_cb_handler()` is invoked-able by a number of events, there is a possibility that when the handler is invoked before TX is completed, and `tx_fifo.altered` is raised. in this case `usbd_cdc_acm` would still enqueue the data from `tx_fifo` to UDC driver at `cdc_acm_tx_fifo_handler()` by calling `usbd_ep_enqueue()`. Since the EP status is still busy at UDC driver side, depending on the enqueue mechanism at UDC, it would be either one of the following scenario:
> > > ```
> > > 1. UDC driver push the request into a queue and returns success, but when the UDC thread tries to execute the TX transaction, it fails and discarded by UDC driver.
> > > 
> > > 2. UDC driver return failure for the calls, and the failure case handling in `cdc_acm_tx_fifo_handler()` discards the buffer:
> > > ```
> > 
> > 
> > For 1), If the driver fails to start a new transfer, the driver would return the buffer to a higher level with an error. For 2), very unlikely, but then there is nothing we can do and your patch does not change it.
> > On which platform in the tree are you observing the problems described in this PR?
> 
> Hi @jfischer-no,
> 
> For 1), there is no way the driver would be able to return the failure status since that the call has returned success when it was queued, and this is the handling from `udc_nrf.c`, it would just return because the endpoint is still busy.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/a51ce53162dd3c313943be8513a829cebdb08938/drivers/usb/udc/udc_nrf.c#L96-L102

Each driver must be able to handle multiple queued transfers, and udc_nrf.c is no different. When a transfer completes or there is an error in handling, the transfer is returned to the higher level
